cli/azd/pkg/commands/pipeline/azdo_provider.go

```diff
@@ -7,65 +7,528 @@ import (
+	repoDetails.remoteUrl = *repo.RemoteUrl
+	repoDetails.repoWebUrl = *repo.WebUrl
+	repoDetails.sshUrl = *repo.SshUrl
+	repoDetails.repoId = repo.Id.String()
+
+	err := p.saveEnvironmentConfig(azdo.AzDoEnvironmentRepoIdName, p.repoDetails.repoId)
+	if err != nil {
+		return fmt.Errorf("error saving repo id to environment %w", err)
+	}
+
+	err = p.saveEnvironmentConfig(azdo.AzDoEnvironmentRepoName, p.repoDetails.repoName)
+	if err != nil {
+		return fmt.Errorf("error saving repo name to environment %w", err)
+	}
+
+	err = p.saveEnvironmentConfig(azdo.AzDoEnvironmentRepoWebUrl, p.repoDetails.repoWebUrl)
+	if err != nil {
+		return fmt.Errorf("error saving repo web url to environment %w", err)
+	}
+
+	return nil
+}
+
+// prompts the user for a new AzDo Git repo and creates the repo
+func (p *AzdoScmProvider) createNewGitRepositoryFromInput(ctx context.Context, console input.Console) (string, error) {
+	connection, err := p.getAzdoConnection(ctx)
+	if err != nil {
+		return "", err
+	}
+
+	var repo *azdoGit.GitRepository
+	for {
+		name, err := console.Prompt(ctx, input.ConsoleOptions{
+			Message:      "Enter the name for your new Azure Devops Repository OR Hit enter to use this name:",
+			DefaultValue: p.repoDetails.projectName,
+		})
+		if err != nil {
+			return "", fmt.Errorf("asking for new project name: %w", err)
+		}
+
+		var message string
+		newRepo, err := azdo.CreateRepository(ctx, p.repoDetails.projectId, name, connection)
+		if err != nil {
+			message = err.Error()
+		}
+		if strings.Contains(message, fmt.Sprintf("A Git repository with the name %s already exists.", name)) {
+			console.Message(ctx, fmt.Sprintf("error: the repo name '%s' is already in use\n", name))
+			continue // try again
+		} else if strings.Contains(message, "TF401025: 'repoName' is not a valid name for a Git repository.") {
+			console.Message(ctx, fmt.Sprintf("error: '%s' is not a valid Azure DevOps repo name. See https://aka.ms/azure-dev/azdo-repo-naming\n", name))
+			continue // try again
+		} else if err != nil {
+			return "", fmt.Errorf("creating repository: %w", err)
+		} else {
+			repo = newRepo
+			break
+		}
+	}
+
+	err = p.StoreRepoDetails(ctx, repo)
+	if err != nil {
+		return "", err
+
+	}
+	return *repo.RemoteUrl, nil
+}
+
+// verifies that a repo exists or prompts the user to select from a list of existing AzDo repos
+func (p *AzdoScmProvider) ensureGitRepositoryExists(ctx context.Context, console input.Console) (string, error) {
+	if p.repoDetails != nil && p.repoDetails.repoName != "" {
+		return p.repoDetails.remoteUrl, nil
+	}
+
+	connection, err := p.getAzdoConnection(ctx)
+	if err != nil {
+		return "", err
+	}
+
+	repo, err := azdo.GetGitRepositoriesInProject(ctx, p.repoDetails.projectName, p.repoDetails.orgName, connection, console)
+	if err != nil {
+		return "", err
+	}
+
+	err = p.StoreRepoDetails(ctx, repo)
+	if err != nil {
+		return "", err
+	}
+
+	remoteParts := strings.Split(p.repoDetails.remoteUrl, "@")
```

Comment:
I wonder if we should use `strings.SplitN` here to ensure we only cut the first `@`?  Or perhaps we should fail when len(remoteParts) != 2? If for some reason there were multiple `@`'s in the remoteUrl I think we'd end up dropping part of the remote?

Reply:
https://github.com/Azure/azure-dev/issues/756